core/types/types.go

```diff
@@ -27,21 +28,76 @@ type Extrinsic []byte
 
 // Block defines a state block
 type Block struct {
-	Header      BlockHeaderWithHash
-	Body        BlockBody
+	Header *BlockHeader
+	Body   *BlockBody
+}
+
+// Block defines a state block
+type BlockWithHash struct {
+	Header      *BlockHeaderWithHash
+	Body        *BlockBody
 	arrivalTime uint64 // arrival time of this block
 }
 
 // GetBlockArrivalTime returns the arrival time for a block
-func (b *Block) GetBlockArrivalTime() uint64 {
+func (b *BlockWithHash) GetBlockArrivalTime() uint64 {
 	return b.arrivalTime
 }
 
 // SetBlockArrivalTime sets the arrival time for a block
-func (b *Block) SetBlockArrivalTime(t uint64) {
+func (b *BlockWithHash) SetBlockArrivalTime(t uint64) {
 	b.arrivalTime = t
 }
 
+// BlockHeader is a state block header
+type BlockHeader struct {
+	ParentHash     common.Hash `json:"parentHash"`
+	Number         *big.Int    `json:"number"`
+	StateRoot      common.Hash `json:"stateRoot"`
+	ExtrinsicsRoot common.Hash `json:"extrinsicsRoot"`
+	Digest         []byte      `json:"digest"` // any additional block info eg. logs, seal
+}
+
+func (bh *BlockHeader) Hash() (common.Hash, error) {
+	enc, err := scale.Encode(bh)
+	if err != nil {
+		return [32]byte{}, err
+	}
+
+	return common.Blake2bHash(enc)
+}
+
+func (bh *BlockHeader) WithHash() (*BlockHeaderWithHash, error) {
+	enc, err := scale.Encode(bh)
+	if err != nil {
+		return nil, err
+	}
+
+	hash, err := common.Blake2bHash(enc)
+	if err != nil {
+		return nil, err
+	}
+
+	return &BlockHeaderWithHash{
+		ParentHash:     bh.ParentHash,
+		Number:         bh.Number,
+		StateRoot:      bh.StateRoot,
+		ExtrinsicsRoot: bh.ExtrinsicsRoot,
+		Digest:         bh.Digest,
+		Hash:           hash,
+	}, nil
+}
+
+func (bh *BlockHeaderWithHash) WithoutHash() *BlockHeader {
+	return &BlockHeader{
+		ParentHash:     bh.ParentHash,
+		Number:         bh.Number,
+		StateRoot:      bh.StateRoot,
+		ExtrinsicsRoot: bh.ExtrinsicsRoot,
+		Digest:         bh.Digest,
+	}
+}
```

Comment:
Really don't love this, let find a better solution (merge if you have to)

Reply:
I also really dislike it :/ I just didn't want to change a bunch of stuff in the block tree and such in this PR. I'll open an issue to deal with it

Reply:
Issue number?

Reply:
https://github.com/ChainSafe/gossamer/issues/506